src/logic/algo/solverWithSteps.ts

```diff
@@ -0,0 +1,162 @@
+  *singleRun(root: TreeNode, hyperParams: MonteCarloTreeSearchHyperParams) {
+    this.kernel = hyperParams.kernel ?? this.kernel;
+
+    const path: TreeNode[] = [];
+    yield* this.select(root, path);
+
+    const leaf = path[path.length - 1];
+    yield* this.expand(leaf);
+
+    const reward = 0;
+    yield* this.simulateNumRolloutsTimes(leaf, hyperParams.numRollout, reward);
+
+    yield* this.backup(path, reward);
+
+    yield;
+  }
+
+  // eslint-disable-next-line
+  *select(node: TreeNode, path: TreeNode[]) {
+    console.log("I'm in select!");
+    console.log(this.nodeRewards);
+    while (true) {
+      path.push(node);
+      if (!this.nodeChildren.has(node)) {
+        break;
+      }
+      if ((this.nodeChildren.get(node) ?? []).length === 0) {
+        break;
+      }
+      let unexploredNodes: Array<TreeNode> = this.nodeChildren.get(node) ?? [];
+      unexploredNodes = unexploredNodes.filter(
+        (unexploredNode) => !this.nodeChildren.has(unexploredNode)
+      );
+      if (unexploredNodes.length > 0) {
+        const unexploredNode: TreeNode | undefined = unexploredNodes.pop();
+        if (unexploredNode) {
+          path.push(unexploredNode);
+        }
+        break;
+      }
+      node = this.uctSelect(node);
+    }
+    yield;
+  }
+
+  choose(node: TreeNode): TreeNode {
+    if (node.children.length === 0) {
+      throw Error("Choose method called on terminal node!");
+    }
+
+    if (!this.nodeChildren.has(node)) {
+      return node.findRandomChild();
+    }
+    return Array.from(this.nodeChildren.get(node) ?? []).reduce((prev, current) =>
+      this.score(prev) > this.score(current) ? prev : current
+    );
+  }
+
+  score(node: TreeNode): number {
+    if (node.children.length === 0) {
+      return Number.NEGATIVE_INFINITY;
+    }
+    return this.nodeRewards.get(node) ?? 0 / (this.nodeVisits.get(node) ?? 1);
+  }
+
+  // eslint-disable-next-line
+  *expand(node: TreeNode) {
+    console.log("I'm in expand!");
+    if (!this.nodeChildren.has(node)) {
+      this.nodeChildren.set(node, node.children);
+    }
+    yield;
+  }
+
+  // eslint-disable-next-line
+  *backup(path: Array<TreeNode>, reward: number) {
+    console.log("I'm in backup!");
+    path.forEach((node) => {
+      const actualVisitsNumber = this.nodeVisits.get(node) ?? 0 + 1;
+      this.nodeVisits.set(node, actualVisitsNumber);
+      const actualReward = this.nodeRewards.get(node) ?? 0 + reward;
+      this.nodeRewards.set(node, actualReward);
+    });
+    yield;
+  }
+
+  simulate(node: TreeNode): number {
+    console.log("I'm in simulate!");
+    while (true) {
+      if (node.children.length === 0) {
+        return node.reward ?? 0;
+      }
+      node = node.findRandomChild();
+    }
+  }
+
+  // eslint-disable-next-line
+  *simulateNumRolloutsTimes(leaf: TreeNode, numRollout: number, reward: number) {
+    for (let i = 0; i < numRollout; i += 1) {
+      reward += this.simulate(leaf);
+    }
+    yield;
```

Comment:
`yield reward` w `for (...)`?
Bo tutaj chcemy mieć możliwość zobaczyć nagrodę którą mamy po każdej rolloucie.